core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -821,9 +858,119 @@ object Parser {
 
   /** Represents where a failure occurred and all the expectations that were broken
     */
-  final case class Error(failedAtOffset: Int, expected: NonEmptyList[Expectation]) {
+  final class ErrorWithInput(
+      val input: String,
+      override val failedAtOffset: Int,
+      override val expected: NonEmptyList[Expectation]
+  ) extends Error(failedAtOffset, expected) {}
+
+  class Error(val failedAtOffset: Int, val expected: NonEmptyList[Expectation])
+      extends Product
+      with Serializable {
     def offsets: NonEmptyList[Int] =
       expected.map(_.offset).distinct
+
+    def copy(
+        failedAtOffset: Int = this.failedAtOffset,
+        expected: NonEmptyList[Expectation] = this.expected
+    ): Error =
+      new Error(failedAtOffset, expected)
+
+    override def productPrefix: String = "Error"
+    def productArity: Int = 2
+    def productElement(x$1: Int): Any = {
+      x$1 match {
+        case 0 => failedAtOffset
+        case 1 => expected
+        case _ => throw new java.lang.IndexOutOfBoundsException(x$1.toString)
+      }
+    }
+
+    override def productIterator: scala.collection.Iterator[Any] =
+      List(
+        failedAtOffset,
+        expected
+      ).iterator
+
+    def canEqual(x$1: scala.Any): Boolean = x$1.isInstanceOf[Error]
+
+    override def hashCode(): scala.Int = {
+      import scala.runtime.Statics
+      var i = -889275714
+      i = Statics.mix(i, productPrefix.hashCode())
+      i = Statics.mix(i, failedAtOffset)
+      i = Statics.mix(i, Statics.anyHash(expected))
+      Statics.finalizeHash(i, 2)
+    }
+    override def toString(): String = s"Error($failedAtOffset, $expected)"
+    override def equals(x$1: Any): Boolean = {
+      if (x$1.isInstanceOf[Error]) {
+        val other = x$1.asInstanceOf[Error]
+        other.failedAtOffset == failedAtOffset &&
+        other.expected == expected
+
+      } else false
+    }
+  }
+
+  object Error extends Serializable {
+    def apply(failedAtOffset: Int, expected: NonEmptyList[Expectation]): Error =
+      new Error(failedAtOffset, expected)
+
+    def unapply(error: Error): Option[(Int, NonEmptyList[Expectation])] =
```

Comment:
if you make the return type `Some` I think the compiler treats this differently, but that might be a mima error.